### PR TITLE
Pass metadata to rule functions (fix #190)

### LINKED ISSUE #190: Pass metadata to rule functions
When #152 lands let's work out how to pass the metadata like add-on type. I think there's going to be quite a lot of it so maybe passing the validation instance might start to make more sense.

### PR DESCRIPTION
**Pull request by tofumatt**
Allows rules to check on the metadata of an add-on, for contextual validation.

### CODE REVIEW

File: src/rules/chromemanifest/categories.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ export const DANGEROUS_CATEGORIES = [
 ];
 
 
-export function checkCategories(triples) {
+export function checkCategories(triples, filename) {
```
Comment (tofumatt) [inline src/rules/chromemanifest/categories.js]: Changing this, even though filename is passed in the `triples` object, so we have a consistent rule function signature for every rule type other than ESLint/JS.

**Comment by tofumatt** [comment]
Alright, have all this working, but I'd love thoughts on if it seems a bit too... sloppy? I dunno, maybe it was just me but I feel like it was a bit of a "cram a bunch of extra stuff" into the scanner to get the tests working.

File: src/scanners/javascript.js [hunk 2]
```diff
@@ -19,22 +21,26 @@ export default class JavaScriptScanner {
     ensureFilenameExists(this.filename);
   }
 
-  scan(_ESLint=ESLint) {
+  scan(_ESLint=ESLint, _rules=this._defaultRules,
```
Comment (muffinresearch) [inline src/scanners/javascript.js]: I think once we get to this it's probably time to consider using an object. Makes it super-obvious from the calling side too.

Comment (tofumatt) [inline src/scanners/javascript.js]: Agreed. I’ll move to that 👍

File: src/rules/chromemanifest/categories.js (hunk 1 above)
Comment (muffinresearch) [inline src/rules/chromemanifest/categories.js]: Makes sense - I :heart: consistency.

File: tests/scanners/test.metadata.js [hunk 3]
```diff
@@ -1,11 +1,24 @@
...
 
+  it('should pass addon metadata to rules', () => {
+    var fakeRules = { fakeRule1: fakeMetadataRuleCheck };
+
+    var metadata = validMetadata();
+    var scanner = new MetadataScanner(metadata, 'fake.xpi', fakeMetadata);
+
+    return scanner.scan(fakeRules)
+      .then((validatorMessages) => {
+        assert.equal(validatorMessages.length, 0);
```
Comment (muffinresearch) [inline tests/scanners/test.metadata.js]: What is this testing wrt passing of the metadata?

Comment (muffinresearch) [inline tests/scanners/test.metadata.js]: Ah I see it's all in the helper.

Comment (muffinresearch) [inline tests/scanners/test.metadata.js]: I wonder if it might be more direct to have this be a sinon stub and check for what was passed in the assertions?

**Comment by muffinresearch** [comment]
Looks good on the whole - I think using a rule stub might make the tests easier to grok at a glance.

**Comment by tofumatt** [comment]
Sounds good, I'll make those tweaks after I'm back from the gym and ask for another quick look. 

- tofumatt

**Comment by tofumatt** [comment]
@muffinresearch This has been tweaked; the CSS Scanner needed a way to load custom rules so I added that for the tests. (I tested it separately from other classes that extend `BaseScanner` because it has its own `scan()` method).

Let me know what you think.

File: src/scanners/css.js [hunk 4]
```diff
@@ -12,7 +12,7 @@ export default class CSSScanner extends BaseScanner {
 
   _defaultRules = rules;
 
-  processCode(cssCode, cssInstruction) {
+  processCode(cssCode, cssInstruction, rules) {
```
Comment (muffinresearch) [inline src/scanners/css.js]: Is this for injection? Might be better to do `_rules=rules` if that's the case to make the distinction.

Comment (tofumatt) [inline src/scanners/css.js]: It is; good call. (I thought that myself but I think I was being lazy; no excuse!)

Comment (muffinresearch) [inline src/scanners/css.js]: Actually should this use _defaultRules like this: https://github.com/mozilla/addons-validator/blob/pass-metadata-to-rules-190/src/scanners/css.js#L51

Comment (tofumatt) [inline src/scanners/css.js]: :+1:

File: tests/scanners/test.css.js [hunk 5]
```diff
@@ -25,6 +25,26 @@ describe('CSSScanner', () => {
...
+    // This rule calls assert.fail() if no metadata is passed to it.
+    sinon.stub(fakeRules, 'metadataPassCheck', metadataPassCheck);
+
+    var scanner = new CSSScanner(code, 'fake.css', {
+      addonMetadata: validMetadata({guid: 'snowflake'}),
+    });
+
+    return scanner.scan(fakeRules)
+      .then((validatorMessages) => {
+        assert.ok(fakeRules.metadataPassCheck.called);
```
Comment (muffinresearch) [inline tests/scanners/test.css.js]: Nice - this is easier to grok.

File: tests/scanners/test.javascript.js [hunk 6]
```diff
@@ -227,6 +228,45 @@ describe('JavaScript Scanner', function() {
...
 
+  it('should pass addon metadata to rules', () => {
+    var fakeRules = { metadata_not_passed: () => {} };
+
+    sinon.stub(fakeRules, 'metadata_not_passed', (context) => {
+      return {
+        Identifier: (node) => {
+          var metadata = context.settings.addonMetadata;
+
+          if (!metadata || typeof metadata.guid === 'undefined') {
```
Comment (muffinresearch) [inline tests/scanners/test.javascript.js]: Is guid guaranteed - it's required but would this mean we'd not be able to carry on with validation even if we wanted to.

Maybe check metadata is an object?

**Comment by muffinresearch** [comment]
r+wc

File: tests/scanners/test.javascript.js (hunk 6 above)
Comment (tofumatt) [inline tests/scanners/test.javascript.js]: GUID is guaranteed here as it's only running this rule in this test, where we pass in a metadata object with a GUID.

I guess I could either check for an object (less explicit) or be even _more_ explicit and test for the exact GUID value. I'll do the latter if that's cool.

Comment (muffinresearch) [inline tests/scanners/test.javascript.js]: Duh, I don't know why but I didn't spot this was a test  - ignore me!

dont-panic-and-carry-on1

Comment (tofumatt) [inline tests/scanners/test.javascript.js]: Lol. Well I cleaned it up a bit anyway ^_^

**Pull request merged**